File: Day10-19/12.py
from collections import defaultdict


def parse_graph(edges):
    graph = defaultdict(list)
    for e in edges:
        vertices = e.split(' ')
        for v in vertices[2:]:
            graph[vertices[0]].append(v.replace(',', '').strip())
    return graph


def bfs_reach(graph, start):
    to_visit = [start]
    reachable = set([start])
    while to_visit:
        curr = to_visit.pop()
        for neighbor in graph[curr]:
            if neighbor not in reachable:
                reachable.add(neighbor)
                to_visit.append(neighbor)
    return reachable


def num_connected_components(graph):
    unmarked = set(graph.keys())
    num_components = 0
    while unmarked:
        component = bfs_reach(graph, unmarked.pop())
        unmarked -= component
        num_components += 1
    return num_components


if __name__ == '__main__':
    with open('12.txt') as f:
        edges = f.readlines()
    graph = parse_graph(edges)
    part1 = bfs_reach(graph, '0')
    print("Part 1: {}".format(len(part1)))
    part2 = num_connected_components(graph)
    print("Part 2: {}".format(part2))

File: Day10-19/12.go
package main

import (
    "io/ioutil"
    "fmt"
    "github.com/golang-collections/collections/stack"
    "strings"
)

func graphReach(graph map[string][]string, start string) []string{
    toVisit := stack.New()
    toVisit.Push(start)
    reachable := make(map[string]bool)
    for ; toVisit.Len() > 0; {
        curr := toVisit.Pop().(string)
        for _, neighbor := range graph[curr] {
            if _, ok := reachable[neighbor]; !ok {
                reachable[neighbor] = true
                toVisit.Push(neighbor)
            }
        }
    }

    keys := []string{}
    for k := range reachable {
        keys = append(keys, k)
    }
    return keys
}

func numConnectedComponents(graph map[string][]string) int {
    marked := make(map[string]bool)
    numComponents := 0
    for vertex := range graph {
        if _, ok := marked[vertex]; ok {
            continue
        }

        component := graphReach(graph, vertex)
        for _, c := range component {
            marked[c] = true
        }
        numComponents++
    }
    return numComponents
}

func parseGraph(edges []string) map[string][]string {
    graph := make(map[string][]string)
    replacer := strings.NewReplacer(",", "", "\n", "")
    for _, e := range edges {
        split := strings.Split(e, " ")
        adjacent := make([]string, len(split) - 2)
        for i, v := range split[2:] {
            adjacent[i] = replacer.Replace(v)
        }
        graph[split[0]] = adjacent
    }
    return graph
}

func main() {
    dat, _ := ioutil.ReadFile("./12.txt")
    graph := parseGraph(strings.Split(string(dat), "\n"))
    part1 := graphReach(graph, "0")
    fmt.Printf("Part 1: %d\n", len(part1))
    part2 := numConnectedComponents(graph)
    fmt.Printf("Part 2: %d\n", part2)
}